streams/src/test/java/org/apache/kafka/streams/integration/MetricsIntegrationTest.java

```diff
@@ -0,0 +1,518 @@
+            testMetricByName(listMetricStore, ALL_LATENCY_MAX, 2);
+            testMetricByName(listMetricStore, RANGE_LATENCY_AVG, 2);
+            testMetricByName(listMetricStore, RANGE_LATENCY_MAX, 2);
+            testMetricByName(listMetricStore, FLUSH_LATENCY_AVG, 2);
+            testMetricByName(listMetricStore, FLUSH_LATENCY_MAX, 2);
+            testMetricByName(listMetricStore, RESTORE_LATENCY_AVG, 2);
+            testMetricByName(listMetricStore, RESTORE_LATENCY_MAX, 2);
+            testMetricByName(listMetricStore, PUT_RATE, 2);
+            testMetricByName(listMetricStore, PUT_TOTAL, 2);
+            testMetricByName(listMetricStore, PUT_IF_ABSENT_RATE, 2);
+            testMetricByName(listMetricStore, PUT_IF_ABSENT_TOTAL, 2);
+            testMetricByName(listMetricStore, GET_RATE, 2);
+            testMetricByName(listMetricStore, DELETE_RATE, 2);
+            testMetricByName(listMetricStore, DELETE_TOTAL, 2);
+            testMetricByName(listMetricStore, PUT_ALL_RATE, 2);
+            testMetricByName(listMetricStore, PUT_ALL_TOTAL, 2);
+            testMetricByName(listMetricStore, ALL_RATE, 2);
+            testMetricByName(listMetricStore, ALL_TOTAL, 2);
+            testMetricByName(listMetricStore, RANGE_RATE, 2);
+            testMetricByName(listMetricStore, RANGE_TOTAL, 2);
+            testMetricByName(listMetricStore, FLUSH_RATE, 2);
+            testMetricByName(listMetricStore, FLUSH_TOTAL, 2);
+            testMetricByName(listMetricStore, RESTORE_RATE, 2);
+            testMetricByName(listMetricStore, RESTORE_TOTAL, 2);
+            return true;
+        } catch (final Throwable e) {
+            return false;
+        }
+    }
+
+    private boolean testCacheMetric() {
+        try {
+            final List<Metric> listMetricCache = new ArrayList<Metric>(kafkaStreams.metrics().values()).stream().filter(m -> m.metricName().group().equals(STREAM_CACHE_NODE_METRICS)).collect(Collectors.toList());
+            testMetricByName(listMetricCache, HIT_RATIO_AVG, 6);
+            testMetricByName(listMetricCache, HIT_RATIO_MIN, 6);
+            testMetricByName(listMetricCache, HIT_RATIO_MAX, 6);
+            return true;
+        } catch (final Throwable e) {
+            return false;
+        }
+    }
+
+    private void testMetricByName(final List<Metric> listMetric, final String metricName, final int numMetric) {
+        try {
+            final List<Metric> metrics = listMetric.stream().filter(m -> m.metricName().name().equals(metricName)).collect(Collectors.toList());
+            Assert.assertEquals("Size of metrics of type:'" + metricName + "' must be equal to:" + numMetric + " but it's equal to " + metrics.size(), numMetric, metrics.size());
+            for (final Metric m : metrics) {
+                Assert.assertNotNull("Metric:'" + m.metricName() + "' must be not null", m.metricValue());
+            }
+        } catch (final Throwable e) {
+            throw e;
```

Comment:
Since in the caller we will capture `throwable` and swallow it by returning false, would the actual error message would be lost?

Reply:
I put a logger, but the checkstyle refuse it, should i find a way to log the error ? Or ignore it ?

Reply:
@khaireddine120 I thought about it again and now I realized adding a logger is not ideal, since it may fail a couple of time within the timeout until it succeeds, but the assert failure would still log an record before succeeds.

So how about doing this?

1) Create an error string variable (initialized as null) which will be passed to `testXX` function, which will also pass it to `testMetricByName` as the additional parameter, then inside `testMetricByName` replace the assertion with the size / not null conditions directly; and if failed modify the passed-in string reference with the necessary information like `which metric name contains unexpected numbers, or name equals to null`.

2) And this string can be passed in as the second parameter if `waitForCondition`, this is because that parameter is actually used as a `Supplier<String>` which is evaluated lazily. So if the condition did not met indeed, that evaluation will then take whatever the current string reference to construct the final message.

WDYT?